### PR TITLE
Emit events from subdirectories when a contained value changes

### LINKED ISSUE #4400: Sub-directories should expose change events
As I was fixing Vltava, I've noticed that we pass directory (root of it) to objects that operate only on sub-directory.
The only reason it's done is because there are no events on sub-directory (IDirectory interface), so consumers need root to be able to install handlers. And have knowledge that they are working with specific sub-directory, when ideally this knowledge should not be at that layer (and be pushed up to the caller).

Fundamentally, the problem is that IDirectory & ISharedDirectory are two district interfaces, while it looks like it should be just one. This shows up in this API returning interface that has no events:
    getSubDirectory(subdirName: string): IDirectory;

Looking at implementation, SubDirectory class does not implement event emitter :(

### PR DESCRIPTION
**Pull request by ChumpChief**
Fixes #4400 

This change adds a "containedValueChanged" event on `IDirectory`s, such that it's no longer necessary to listen to "valueChanged" on the `ISharedDirectory` and inspect the directory path.

### CODE REVIEW

File: packages/dds/map/src/interfaces.ts [hunk 1]
```diff
@@ -191,18 +190,36 @@ export interface IDirectory extends Map<string, any> {
...
+        target: IEventThisPlaceHolder,
+    ) => void);
+}
+
+export interface IDirectoryEvents extends IEvent {
+    (event: "containedValueChanged", listener: (
+        changed: IValueChanged,
+        local: boolean,
+        op: ISequencedDocumentMessage,
+        target: IEventThisPlaceHolder,
```
Comment (anthony-murphy) [inline packages/dds/map/src/interfaces.ts:215]: can we stop progagating the op. i think it just creates more problems as people think they can just use the values as is, but thats not really the case, and not sure what the uses case is

Comment (ChumpChief) [reply packages/dds/map/src/interfaces.ts:215]: Good point

File: packages/dds/map/src/interfaces.ts [hunk 2]
```diff
@@ -191,18 +190,36 @@ export interface IDirectory extends Map<string, any> {
...
+    ) => void);
 }
 
 /**
  * Interface describing a shared directory.
  */
-export interface ISharedDirectory extends ISharedObject<ISharedDirectoryEvents>, IDirectory {
+export interface ISharedDirectory extends ISharedObject<ISharedDirectoryEvents>, IDirectoryNoEvents {
```
Comment (anthony-murphy) [inline packages/dds/map/src/interfaces.ts:222]: this is all a bit wonky. I think we should get ride of IDirectoryNotEvents, and here you can just remove the events from IDirectory,  Omit<Omit<Omit<IDirectory,"on">,"once">,"off">. I'm also unlcear why shared directory does have containedValueChanged, shouldn't we get that on the root as well?

Comment (ChumpChief) [reply packages/dds/map/src/interfaces.ts:222]: Yeah I'm fighting with the typing here -- I'll try the Omit approach and see how that does.

Shared directory should have containedValueChanged -- I had ISharedDirectoryEvents extending IDirectoryEvents at one point, think I just accidentally removed that.

File: packages/dds/map/src/directory.ts [hunk 3]
```diff
@@ -428,24 +430,31 @@ export class SharedDirectory extends SharedObject<ISharedDirectoryEvents> implem
         super(id, runtime, attributes);
         this.localValueMaker = new LocalValueMaker(this.serializer);
         this.setMessageHandlers();
+        // Mirror the containedValueChanged op on the SharedDirectory
+        this.root.on(
+            "containedValueChanged",
+            (changed: IValueChanged, local: boolean, op: ISequencedDocumentMessage) => {
+                this.emit("containedValueChanged", changed, local, op);
```
Comment (anthony-murphy) [inline packages/dds/map/src/directory.ts:437]: >containedValueChanged [](start = 27, length = 21)

as i call out on the interfaces file, this event isn't defined on ISharedDirectory, so i don't think it is part of the typing here

File: packages/dds/map/src/interfaces.ts (hunk 2 above)
Comment (anthony-murphy) [reply packages/dds/map/src/interfaces.ts:222]: yeah. it's tricky. you can do something like SharedObject<ISharedDirectoryEvents & IDirectoryEvents>, there is also the ExtendEventProvider type in common utils, but i'm not sure what fits with the composition you are doing. happy to help,

---
In reply to: [540356177](https://github.com/microsoft/FluidFramework/pull/4556#discussion_r540356177) [](ancestors = 540356177)

Comment (anthony-murphy) [reply packages/dds/map/src/interfaces.ts:222]: if i understand correctly. you don't want shared object events on IDirectory, and you want all events on ISharedDirectory. For that i think omiting is the way. you should omit from IDirectory on ISharedDirectory, and intersect(&) the events types in the ISharedObject generic. I think that will work, and doesn't require extra types

---
In reply to: [540358116](https://github.com/microsoft/FluidFramework/pull/4556#discussion_r540358116) [](ancestors = 540358116,540356177)

File: packages/dds/map/src/interfaces.ts (hunk 1 above)
Comment (ChumpChief) [reply packages/dds/map/src/interfaces.ts:207]: Removed in latest

File: packages/dds/map/src/interfaces.ts (hunk 2 above)
Comment (ChumpChief) [reply packages/dds/map/src/interfaces.ts]: Latest iteration uses Omit but then adds back the Symbols that get stripped out, which allows us to stop exporting an intermediate no-event interface.  LMK if there are other things you'd like me to try.

File: packages/dds/map/src/directory.ts (hunk 3 above)
Comment (ChumpChief) [reply packages/dds/map/src/directory.ts]: Fixed in latest

**Review by anthony-murphy — APPROVED** [review]

**Pull request merged**